### ISSUE TITLE
SIGSEGV in vim at ex_eval.c:720

### ISSUE THREAD

**Issue by irfanariq** [vim/vim#8954]
Hello,

We are currently working on fuzz testing feature, and we found a **SIGSEGV** on `vim`.

The stack traces are as follow:
```st
Program received signal SIGSEGV, Segmentation fault.
0x00005555558125f5 in finish_exception (excp=0x604000059c10) at ex_eval.c:720
720	    caught_stack = caught_stack->caught;
#0  0x00005555558125f5 in finish_exception (excp=0x604000059c10) at ex_eval.c:720
#1  0x000055555581b6cd in cleanup_conditionals (cstack=0x7fffffffd270, searched_cond=0, inclusive=0) at ex_eval.c:2404
#2  0x0000555555817806 in do_throw (cstack=0x7fffffffd270) at ex_eval.c:1550
#3  0x0000555555810195 in do_errthrow (cstack=0x7fffffffd270, cmdname=0x555555de57e0 "finally") at ex_eval.c:361
#4  0x00005555557e9409 in do_one_cmd (cmdlinep=0x7fffffffd110, flags=7, cstack=0x7fffffffd270, fgetline=0x5555557e254c <get_loop_line>, cookie=0x7fffffffd1d0) at ex_docmd.c:2667
#5  0x00005555557dfb3e in do_cmdline (cmdline=0x611000131b80 "localo2 p\" vim-plug: Vim plugin mnum_plugs = nu", '=' <repeats 23 times>, fgetline=0x555555af369e <getsourceline>, cookie=0x7fffffffdb30, flags=7) at ex_docmd.c:1000
#6  0x0000555555af1b76 in do_source (fname=0x7fffffffe140 "crashlogs2_faster11_friend_vim_power-only-210914_2__crashes_id:000047,sig:11,src:006992,argv:000478,time:70513139,op:havoc,rep:8_input", check_other=0, is_vimrc=0, ret_sid=0x0) at scriptfile.c:1406
#7  0x0000555555da6129 in source_startup_scripts (parmp=0x555555ff13c0 <params>) at main.c:3144
#8  0x0000555555d9e908 in main (argc=20, argv=0x7fffffffdd48) at main.c:413
```
Here is the full stack trace : [full_stacktrace_poc_2.zip](https://github.com/vim/vim/files/7279457/full_stacktrace_poc_2.zip)

**Step to reproduce**

We configured `vim` using `CFLAGS="-g -O0 -fsanitize=address" CXXFLAGS="-g -O0 -fsanitize=address" LDFLAGS="-g -O0 -fsanitize=address" ./configure --prefix=$(pwd)/` and build it using `make -j10`, and run it with:

```
./vim -o1 -b -S -y --noplugin -O2 -E -i --startuptime -A --ttyfail NONE -u <attached file> -S -R -o2 + -V1
```
Attachment: [input_vim_poc_2.zip](https://github.com/vim/vim/files/7279458/input_vim_poc_2.zip)

**Environment**
- OS: Ubuntu 18.04.5 LTS
- GCC version: gcc 7.5.0
- vim version: latest commit (50c56893) of master branch in github ([link]([https://github.com/vim/vim/tree/50c56893423eb6ad2154a4151e67f7097f52efb6]))

Thank you.

**Comment by dpelle** [comment]
I can reproduce the bug with vim-8.2.3469 (latest).
I minimized the reproduction to this small script:
```
$ cat >bug.vim <<EOF
try
  if x | endif
catch
  for l in []
finally
EOF
```
```
$ ./vim --clean -S bug.vim
Vim: Caught deadly signal SEGV
Vim: Finished.
```
Of course the input script is errors, but it should not cause a crash.
Valgrind or asan don't give more info.
Stack in gdb:
```
$ gdb --args ./vim --clean -S bug.vim
(gdb) run
(gdb) bt
#0  0x00000000004aee95 in finish_exception (excp=0xc5d660) at ex_eval.c:720
#1  0x00000000004ad1b7 in cleanup_conditionals (cstack=0x7fffffffca20, searched_cond=0, inclusive=0) at ex_eval.c:2404
#2  0x00000000004ab471 in do_throw (cstack=0x7fffffffca20) at ex_eval.c:1550
#3  0x00000000004ab0b5 in do_errthrow (cstack=0x7fffffffca20, cmdname=0x724d3b "finally") at ex_eval.c:361
#4  0x000000000049c7fd in do_one_cmd (cmdlinep=0x7fffffffd068, flags=7, cstack=0x7fffffffca20, fgetline=0x49a4e0 <get_loop_line>, cookie=0x7fffffffc960) at ex_docmd.c:2667
#5  0x00000000004996b0 in do_cmdline (cmdline=0x9db4c0 "try", fgetline=0x5ef9d0 <getsourceline>, cookie=0x7fffffffd190, flags=7) at ex_docmd.c:1000
#6  0x00000000005ef36f in do_source (fname=0x9e03e3 "bug.vim", check_other=0, is_vimrc=0, ret_sid=0x0) at scriptfile.c:1406
#7  0x00000000005ee7b6 in cmd_source (fname=0x9e03e3 "bug.vim", eap=0x7fffffffd3e8) at scriptfile.c:971
#8  0x00000000005ee6cc in ex_source (eap=0x7fffffffd3e8) at scriptfile.c:997
#9  0x000000000049c578 in do_one_cmd (cmdlinep=0x7fffffffdc88, flags=11, cstack=0x7fffffffd640, fgetline=0x0, cookie=0x0) at ex_docmd.c:2611
#10 0x00000000004996b0 in do_cmdline (cmdline=0x9e03a0 "so bug.vim", fgetline=0x0, cookie=0x0, flags=11) at ex_docmd.c:1000
#11 0x000000000049a2e1 in do_cmdline_cmd (cmd=0x9e03a0 "so bug.vim") at ex_docmd.c:594
#12 0x000000000070eae3 in exe_commands (parmp=0x9c8858 <params>) at main.c:3081
#13 0x000000000070db95 in vim_main2 () at main.c:773
#14 0x000000000070b5d8 in main (argc=4, argv=0x7fffffffde58) at main.c:425
```
```
 712│ /*
 713│  * Remove an exception from the caught stack.
 714│  */
 715│     static void
 716│ finish_exception(except_T *excp)
 717│ {
 718│     if (excp != caught_stack)
 719│         internal_error("finish_exception()");
 720├>    caught_stack = caught_stack->caught;
```
```
(gdb) p caught_stack
$1 = (except_T *) 0x0
```
So we access `NULL` at line `ex_eval.c:720`.

**Comment by irfanariq** [comment]
Thank you for your response.

I have checked whether the bug is already fixed or not on the latest commit of the master branch ([6d19798](https://github.com/vim/vim/tree/6d197987745427801a2195836ef89e48537afbeb)) and it seems the bug is still occurs

**Comment by dpelle** [comment]
@ irfanariq  wrote:

> I have checked whether the bug is already fixed or not on the latest
> commit of the master branch (6d19798) and it seems the bug is still occurs

I confirm.
I can also still reproduce this issue using the same steps with the latest
vim-8.2.3477, yet the bug was meant to be fixed in vim-8.2.3470.

```
$ cat >bug.vim <<EOF
try
  if x | endif
catch
  for l in []
finally
EOF
```
```
$  gdb --args ./vim --clean -S bug.vim
(gdb) run
Program received signal SIGSEGV, Segmentation fault.
(gdb) bt
#0  0x00000000004aef95 in finish_exception (excp=0xc5abf0) at ex_eval.c:720
#1  0x00000000004ad2be in cleanup_conditionals (cstack=0x7fffffffca20, searched_cond=0, inclusive=0) at ex_eval.c:2405
#2  0x00000000004ab561 in do_throw (cstack=0x7fffffffca20) at ex_eval.c:1550
#3  0x00000000004ab1a5 in do_errthrow (cstack=0x7fffffffca20, cmdname=0x724e4b "finally") at ex_eval.c:361
#4  0x000000000049c8ea in do_one_cmd (cmdlinep=0x7fffffffd068, flags=7, cstack=0x7fffffffca20, fgetline=0x49a5c0 <get_loop_line>, 
    cookie=0x7fffffffc960) at ex_docmd.c:2670
#5  0x0000000000499790 in do_cmdline (cmdline=0x9db5b0 "try", fgetline=0x5efae0 <getsourceline>, cookie=0x7fffffffd190, flags=7)
    at ex_docmd.c:1000
#6  0x00000000005ef47f in do_source (fname=0x9e03e3 "bug.vim", check_other=0, is_vimrc=0, ret_sid=0x0) at scriptfile.c:1406
#7  0x00000000005ee8c6 in cmd_source (fname=0x9e03e3 "bug.vim", eap=0x7fffffffd3e8) at scriptfile.c:971
#8  0x00000000005ee7dc in ex_source (eap=0x7fffffffd3e8) at scriptfile.c:997
#9  0x000000000049c665 in do_one_cmd (cmdlinep=0x7fffffffdc88, flags=11, cstack=0x7fffffffd640, fgetline=0x0, cookie=0x0) at ex_docmd.c:2614
#10 0x0000000000499790 in do_cmdline (cmdline=0x9e03a0 "so bug.vim", fgetline=0x0, cookie=0x0, flags=11) at ex_docmd.c:1000
#11 0x000000000049a3c1 in do_cmdline_cmd (cmd=0x9e03a0 "so bug.vim") at ex_docmd.c:594
#12 0x000000000070ebf3 in exe_commands (parmp=0x9c8878 <params>) at main.c:3081
#13 0x000000000070dca5 in vim_main2 () at main.c:773
#14 0x000000000070b6e8 in main (argc=4, argv=0x7fffffffde58) at main.c:425
```

A little bit simpler, this also crashes:
```
$ cat >bug.vim <<EOF
try
  echo x
catch
  for l in []
finally
EOF
```
```
$  ./vim --clean -S bug.vim
Vim: Caught deadly signal SEGV
Vim: Finished.
Segmentation fault (core dumped)
```

Running `make test_trycatch` did no crash.  So the added test
in patch 8.2.3470 was not sufficient to detect the issue.  I can try to
come up with a test that reproduce it but I may not find time today.

**Comment by brammool** [comment]
> @ irfanariq  wrote:
> 
> > I have checked whether the bug is already fixed or not on the latest
> > commit of the master branch (6d19798) and it seems the bug is still occurs
> 
> I confirm.
> I can also still reproduce this issue using the same steps with the latest
> vim-8.2.3477, yet the bug was meant to be fixed in vim-8.2.3470.
> 
> ```
> $ cat >bug.vim <<EOF
> try
>   if x | endif
> catch
>   for l in []
> finally
> EOF
> ```
> ```
> $  gdb --args ./vim --clean -S bug.vim
> (gdb) run
> Program received signal SIGSEGV, Segmentation fault.
> (gdb) bt
> #0  0x00000000004aef95 in finish_exception (excp=0xc5abf0) at ex_eval.c:720

[...]

> A little bit simpler, this also crashes:
> ```
> $ cat >bug.vim <<EOF
> try
>   echo x
> catch
>   for l in []
> finally
> EOF
> ```
> ```
> $  ./vim --clean -S bug.vim
> Vim: Caught deadly signal SEGV
> Vim: Finished.
> Segmentation fault (core dumped)
> ```
> 
> Running `make test_trycatch` did no crash.  So the added test
> in patch 8.2.3470 was not sufficient to detect the issue.  I can try to
> come up with a test that reproduce it but I may not find time today.

I hope you can.  I had the problem fixed, wrote the test, then noticed
another test failed, so I adjusted the code.  But it looks like using
assert_fails() changes what happens, the test doesn't reproduce the
original problem.

I will send a patch to fix the problem, please suggest a test that fails
without this patch.

-- 
hundred-and-one symptoms of being an internet addict:
97. Your mother tells you to remember something, and you look for
    a File/Save command.

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by dpelle** [comment]
> > I can try to come up with a test that reproduce it but I may not find time today.
>
> I hope you can.  I had the problem fixed, wrote the test, then noticed
> another test failed, so I adjusted the code. 

I changed the test so that it now triggers the crash when using vim
without the fix. See this PR:

https://github.com/vim/vim/pull/8961
